content/blog/2024/proxy-store.md

```diff
@@ -0,0 +1,174 @@
+// "Setting input to Hello!"
+```
+
+Every time we set the value of `state.input`, we log the prop and value. Note
+that we use the proxied object `state`, and not the `initialState` object.
+Because we are trapping `set`, we need to set the value with `Reflect.set()`. We
+also need to return a Boolean specifying whether the `set` was successful.
+
+## A more complex Proxy state
+
+```js
+const initialState = {
+  inputFormat: 'scss',
+  outputFormat: 'expanded',
+  inputValue: '',
+  compilerHasError: false,
+};
+
+const playgroundState = new Proxy(initialState, {
+  set(state, prop, ...rest) {
+    // Set state first so called functions have access
+    const set = Reflect.set(state, prop, ...rest);
+    if (['inputFormat', 'outputFormat', 'inputValue'].includes(prop)) {
+      updateCSS();
+      updateURL();
+    } else if (prop === 'compilerHasError') {
+      updateErrorState();
+    }
+    return set;
+  },
+});
+```
+
+{% callout %}
+This is a bit simplified from the [full source code](https://github.com/sass/sass-site/blob/main/source/assets/js/playground.ts).
+{% endcallout %}
+
+With this, we've achieved a few things. First, we have a central location for
+the data we need to access. Second, we have a central place to declare what
+happens when a value changes.
+
+When a user changes the input format, we simply need to update the state with
+`playgroundState.inputFormat = 'sass';` and `updateCSS` and `updateURL` are
+automatically called. We no longer need to remember to trigger the side effects
+in each place where state is updated.
+
+## Getting ahead of ourselves
+
+For this particular project, I only needed to trigger side effects, so I only
+needed to use the `set` trap. One thing I'm interested in trying out in a future
+project is using a `get` trap to set up a computed value based on other values
+in the state.
+
+```js
+const initialState = { input: '' };
+
+const state = new Proxy(initialState, {
+  get(target, prop, receiver) {
+    if (prop === 'isEmpty') {
+      return target.input.length === 0;
+    }
+    return Reflect.get(...arguments);
+  },
+});
+
+console.log(state.isEmpty); // true
+state.input = 'has a value';
+console.log(state.isEmpty); // false
+```
+
+This pattern would be useful when you need to access a value derived from other
+pieces of the state at a different point than when you set the state.
+
+## When we would not want to use a Proxy store
+
+This isn't a universal solution. If you have an existing library, use it. If you
+find yourself abstracting out things like `watch` or `computed` for your proxy
+state, you are starting down the road to developing your own framework, and it
+might be a good time to pause and see if your application has grown complex
+enough to bring in something more robust.
+
+This also works best when your logic is contained within a single file. If
+you're wanting to share state values across files, you'll likely be happier
+reaching for another solution.
```

Comment:
Hm, I'm curious about this advice -- maybe because I initially read it as implying that the Proxy should only be used in a single file where it's defined. I could imagine an app with relatively simple state logic using a Proxy that's defined in e.g. `store.ts` but then interacted with in various places throughout the app.

Reply:
Good point. I updated to clarify that while you can easily set and access the state across files, the side effects are best colocated with the store itself. 

You _could_ add a `addSideEffect(prop, callback)` method to the store, and then store and call those, but I'd probably recommend against it.